arangod/Pregel/GraphStore.cpp

```diff
@@ -85,41 +85,72 @@ GraphStore<V, E>::~GraphStore() {
 }
 
 template <typename V, typename E>
-std::unordered_map<ShardID, uint64_t> GraphStore<V, E>::_preallocateMemory() {
+std::map<CollectionID, std::vector<VertexShardInfo>>
+  GraphStore<V, E>::_allocateSpace() {
   if (_vertexData || _edges) {
     THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL,
                                    "Only allocate messages once");
     //_edges->resize(count);
   }
 
-  double t = TRI_microtime();
-  std::unique_ptr<transaction::Methods> countTrx(_createTransaction());
-  std::unordered_map<ShardID, uint64_t> shardSizes;
+  std::map<CollectionID, std::vector<VertexShardInfo>> result;
+  
   LOG_TOPIC(DEBUG, Logger::PREGEL) << "Allocating memory";
   uint64_t totalMemory = TRI_totalSystemMemory();
-
+  
+  std::map<CollectionID, std::vector<ShardID>> const& vertexCollMap =
+  _config->vertexCollectionShards();
+  std::map<CollectionID, std::vector<ShardID>> const& edgeCollMap =
+  _config->edgeCollectionShards();
+  
   // Allocating some memory
   uint64_t vCount = 0;
-  for (auto const& shard : _config->localVertexShardIDs()) {
-    OperationResult opResult = countTrx->count(shard, transaction::CountType::Normal);
-    if (opResult.fail() || _destroyed) {
-      THROW_ARANGO_EXCEPTION(TRI_ERROR_BAD_PARAMETER);
-    }
-    shardSizes[shard] = opResult.slice().getUInt();
-    vCount += opResult.slice().getUInt();
-  }
-  _index.resize(vCount);
-  
   uint64_t eCount = 0;
-  for (auto const& shard : _config->localEdgeShardIDs()) {
-    OperationResult opResult = countTrx->count(shard, transaction::CountType::Normal);
-    if (opResult.fail() || _destroyed) {
-      THROW_ARANGO_EXCEPTION(TRI_ERROR_BAD_PARAMETER);
+  for (auto const& pair : vertexCollMap) {
+    std::vector<ShardID> const& vertexShards = pair.second;
+    for (size_t i = 0; i < vertexShards.size(); i++) {
+
+      VertexShardInfo info;
+      info.vertexShard = vertexShards[i];
+      info.trx = _createTransaction();
```

Comment:
I think `_createTransaction()` does not call `trx->pinData(cid);`, as the previous transaction creation functionality did. Can the call to `pinData` be added so it's safe for MMFiles?